[mono] Make some API functions public and private

lambdageek opened the pull request:
See https://github.com/dotnet/runtime/pull/33633 and https://github.com/xamarin/xamarin-macios/blob/master/runtime/exports.t4

* `mono_install_load_aot_data_hook` was already a `MONO_API`, but it was not in a public header.
  It is added to `mono/jit/mono-private-unstable.h`
* `mono_gc_init_finalizer_thread` was [used by Xamarin.iOS](https://github.com/xamarin/xamarin-macios/blob/63ab48e679716264b33af00f86c24426da1d9f97/runtime/mono-runtime.h.t4#L278) for a long time without being in any public header.  It's one of the reasons we have to compile Mono for XI with `--disable-visibility-hidden`.  Make the function a proper `MONO_API` unconditionally - but make it do nothing if the runtime is compiled without `--with-lazy-thread-creation` (the default).  It is now a public Mono API function.
* `mono_trace_init` was a `MONO_API` function in a non-public header that was already used by embedders that need to set up logger hooks.  It is now not necessary to call this before calling the logger functions to set up logger hooks such as `mono_trace_set_log_handler`

lambdageek committed:
[mini] Move mono_install_load_aot_data_hook to a public header

It's already a MONO_API and it is used, for example, by Xamarin.iOS, but it
wasn't in a public header.

Mark it external only.  There are no uses of it inside the runtiem

lambdageek committed:
[gc] Make mono_gc_init_finalizer_thread a public API

It was already used by Xamarin.iOS but wasn't in a public header and required
compiling with --disable-visibility-hidden in order to dlsym the symbol.

If --with-lazy-gc-thread-creation is used, embedders must call this function to
create the finalizer therad.  Otherwise the function does nothing and the
runtime will create the finalizer thread automatically.

lambdageek committed:
[utils] Move mono_trace_init to a public header

It was already marked as MONO_API, but it was not in a public header.
It is used by embedders such as Xamarin.iOS

EgorBo commented:
heh, I was about to file an issue 🙂

EgorBo commented:
since https://github.com/dotnet/runtime/pull/33633 is already merged could you please update `runtime.m` ? https://github.com/dotnet/runtime/blob/master/src/mono/netcore/sample/iOS/runtime.m#L15-L20

lambdageek committed:
Update ios sample

lambdageek commented:
Verified that the ios sample builds in Xcode.

vargaz commented:
So we didn't put them in public headers to avoid having to avoid users depending on them and thus having to support them forever.

akoeplinger commented:
@vargaz you could argue that with XI relying on them we're already on that path. Do you see any of these APIs as particularly problematic?

lambdageek commented:
new plan:

1. `mono_trace_init` is going to stay private. instead i'm going to update `mono_trace_set_log_handler` and other logging APIs to implicitly initialize the logger.
2. `mono_gc_init_finalizer_thread` will become a public API (there's no other way to work with a `--with-lazy-thread-creation` runtime)
3. Move `mono_install_load_aot_data_hook` to the new `mono/jit/mono-private-unstable.h` header from #33869

lambdageek committed:
Revert "[utils] Move mono_trace_init to a public header"

This reverts commit a0cc08768e97c7b630ccbf54e48020b06436f874.

lambdageek committed:
[utils] init mono_logger in all public API functions

Embedders don't need to call mono_trace_init before calling
mono_trace_set_log_handler, mono_trace_set_print_handler or mono_trace_set_printerr_handler

lambdageek committed:
[jit] Move mono_install_load_aot_data_hook to mono-private-unstable.h

We do not guarantee that this API will be stable

lambdageek commented:
Changed this PR to make `mono_install_load_aot_data_hook` an unstable API, and to make `mono_trace_init` unnecessary.

@EgorBo could you check that the sample still compiles.

EgorBo commented:
@lambdageek just checked, compiles and works 👍
